### PR TITLE
[flink] Make DPP distribute the splits fairly for flink

### PR DESCRIPTION
**Pull request by HuangXingBo**
### Purpose
 
The existing DPP assigner implementation first assigns fairly and then assigns according to DPP, which may result in uneven splits for different tasks.

### CODE REVIEW

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/assigners/DynamicPartitionPruningPreAssignSplitAssigner.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+import java.util.Collection;
+import java.util.stream.Collectors;
+
+/**
+ * Pre-calculate which splits each task should process by given {@link DynamicFilteringData}, and
+ * then distribute the splits fairly.
+ */
+public class DynamicPartitionPruningPreAssignSplitAssigner extends PreAssignSplitAssigner {
+    public DynamicPartitionPruningPreAssignSplitAssigner(
```
Comment (wwj6591812) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/assigners/DynamicPartitionPruningPreAssignSplitAssigner.java:37]: add a blank line

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/assigners/DynamicPartitionPruningPreAssignSplitAssigner.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            DynamicFilteringData dynamicFilteringData) {
+        super(
+                splitBatchSize,
+                parallelism,
+                splits.stream()
+                        .filter(s -> filter(partitionRowProjection, dynamicFilteringData, s))
+                        .collect(Collectors.toList()));
+    }
+
+    private static boolean filter(
```
Comment (wwj6591812) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/assigners/DynamicPartitionPruningPreAssignSplitAssigner.java:51]: why static？

Comment (wwj6591812) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/assigners/DynamicPartitionPruningPreAssignSplitAssigner.java:51]: Change DynamicPartitionPruningAssigner#filter to public static, and then delete DynamicPartitionPruningPreAssignSplitAssigner#filter?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/assigners/DynamicPartitionPruningPreAssignSplitAssigner.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import org.apache.flink.table.connector.source.DynamicFilteringData;
+
+import java.util.Collection;
+import java.util.stream.Collectors;
+
+/**
+ * Pre-calculate which splits each task should process by given {@link DynamicFilteringData}, and
+ * then distribute the splits fairly.
+ */
+public class DynamicPartitionPruningPreAssignSplitAssigner extends PreAssignSplitAssigner {
```
Comment (wwj6591812) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/assigners/DynamicPartitionPruningPreAssignSplitAssigner.java:36]: May be this class doesn't need.
You could move filter logic to PreAssignSplitAssigner.

**Review by yuzelin — APPROVED** [review]
+1

**Pull request merged**